### PR TITLE
Add xeno abilities

### LINKED ISSUE #388: Punch

### LINKED ISSUE #389: Lunge

### LINKED ISSUE #390: Fling

### LINKED ISSUE #402: Charge

### LINKED ISSUE #410: Stomp

### LINKED ISSUE #414: Gut

### LINKED ISSUE #419: Screech

### PR DESCRIPTION
**Pull request by MetalSage**
## About the PR

Fixes #388 
Fixes #390 
Fixes #389 
Fixes #402 
Fixes #410 
Fixes #414 
Fixes #419 

- [x] Added warrior punch action
- [x] Added warrior fling action
- [x] Added warrior lunge action
- [x] Added crusher charge action
- [x] Added crusher stomp action
- [x] Added queen gut action
- [x] Added queen screech action
- [x] Defender headbutt action now requires an entity to use

## Media

https://github.com/CM-14/CM-14/assets/74924875/caeb8463-4963-4380-917d-8346d5a94f37

https://github.com/CM-14/CM-14/assets/74924875/f85f2089-2f40-40a6-a752-416328f7f08c

https://github.com/CM-14/CM-14/assets/74924875/c502c7e3-d196-4e3f-8e16-4ca942895efd

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl:
- add: Added the Warrior's punch, fling and lunge actions.
- add: Added the Crusher's charge and stomp actions.
- add: Added the Queen's gut and screech actions.
- fix: Defender headbutt action now requires a marine target to be used.

### CODE REVIEW

**Comment by Skarletto** [comment]
My only feedback would be to add the visual cue for the crusher's stomp, and give it a charge time on the charge ability as seen on cm13 (the charge isn't instant)

https://github.com/CM-14/CM-14/assets/122584947/67d14895-0cab-43e5-8bc1-4ef787a02899

**Comment by MetalSage** [comment]
> My only feedback would be to add the visual cue for the crusher's stomp, and give it a charge time on the charge ability as seen on cm13 (the charge isn't instant), but this is great already
> 
>  Video.sans.titre.Realisee.avec.Clipchamp.35.1.mp4

fixed

https://github.com/CM-14/CM-14/assets/74924875/0acb8a96-f776-4988-a0ea-94a754965a0d

File: Content.Server/_CM14/Xenos/Gut/GutSystem.cs [hunk 1]
```diff
@@ -0,0 +1,5 @@
+﻿using Content.Shared._CM14.Xenos.Gut;
+
+namespace Content.Server._CM14.Gut;
+
+public sealed class GutSystem : SharedXenoGutSystem;
```
Comment (DrSmugleaf) [inline Content.Server/_CM14/Xenos/Gut/GutSystem.cs:5]: Same here.

File: Content.Server/_CM14/Xenos/Punch/PunchSystem.cs [hunk 2]
```diff
@@ -0,0 +1,5 @@
+﻿using Content.Shared._CM14.Xenos.Punch;
+
+namespace Content.Server._CM14.Punch;
+
+public sealed class PunchSystem : SharedXenoPunchSystem;
```
Comment (DrSmugleaf) [inline Content.Server/_CM14/Xenos/Punch/PunchSystem.cs:5]: Same here.

File: Content.Shared/_CM14/Xenos/Gut/XenoGutDoAfterEvent.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+namespace Content.Shared._CM14.Xenos.Gut;
+
+[Serializable, NetSerializable]
+public sealed partial class XenoGutDoAfterEvent : DoAfterEvent
+{
+    public override DoAfterEvent Clone()
+    {
+        return this;
+    }
+}
```
Comment (DrSmugleaf) [inline Content.Shared/_CM14/Xenos/Gut/XenoGutDoAfterEvent.cs:14]: In cases where `Clone()` just returns `this`, you can make the event inherit from `SimpleDoAfterEvent` instead.
```suggestion
public sealed partial class XenoGutDoAfterEvent : SimpleDoAfterEvent;
```

File: Content.Shared/_CM14/Xenos/Charge/XenoChargeDoAfterEvent.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    public XenoChargeDoAfterEvent(NetCoordinates coordinates)
+    {
+        Coordinates = coordinates;
+    }
+
+    public override DoAfterEvent Clone()
+    {
+        return this;
+    }
+}
```
Comment (DrSmugleaf) [inline Content.Shared/_CM14/Xenos/Charge/XenoChargeDoAfterEvent.cs:22]: In cases where `Clone()` just returns `this`, you can make the event inherit from `SimpleDoAfterEvent` instead.
```suggestion
public sealed partial class XenoChargeDoAfterEvent : SimpleDoAfterEvent
{
    [DataField]
    public NetCoordinates Coordinates;

    public XenoChargeDoAfterEvent(NetCoordinates coordinates)
    {
        Coordinates = coordinates;
    }
}
```

File: Content.Shared/_CM14/Xenos/Charge/SharedXenoChargeSystem.cs [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        var length = diff.Length();
+        diff *= xeno.Comp.Range / 3 / length;
+
+        _stun.TryParalyze(targetId, xeno.Comp.StunTime, true);
+        _throwing.TryThrow(targetId, diff, 10);
+
+        if (_net.IsServer &&
+            HasComp<MarineComponent>(targetId))
+            SpawnAttachedTo(xeno.Comp.Effect, targetId.ToCoordinates());
+    }
```
Comment (DrSmugleaf) [inline Content.Shared/_CM14/Xenos/Charge/SharedXenoChargeSystem.cs:136]: The code here is almost the same as the one in HeadbuttSystem, this code should be in a separate method that both call.

File: Resources/Prototypes/_CM14/Actions/Xeno/xeno_offense_actions.yml [hunk 6]
```diff
@@ -98,15 +98,14 @@
   name: Headbutt (10) # TODO CM14 proper plasma costs
   description: Charge in a direction and push back any enemies in your way.
   components:
-  - type: WorldTargetAction
+  - type: EntityTargetAction
```
Comment (DrSmugleaf) [inline Resources/Prototypes/_CM14/Actions/Xeno/xeno_offense_actions.yml:101]: Would this not make the defender always undershoot if they try to headbutt a fleeing marine?

**Review by DrSmugleaf — CHANGES_REQUESTED** [review]

File: Resources/Prototypes/_CM14/Actions/Xeno/xeno_offense_actions.yml (hunk 6 above)
Comment (MetalSage) [reply Resources/Prototypes/_CM14/Actions/Xeno/xeno_offense_actions.yml:101]: I've tested this and I think it works quite well, the headbutt has a small margin that is enough to catch up to a marine
https://youtu.be/Yx6fHGWab_U

**Comment by MetalSage** [comment]
Made the suggested changes, but I was unable to make a generic method for Headbutt and Charge, stuck on an error:
```
[FATL] unhandled: Robust.Shared.IoC.Exceptions.UnregisteredDependencyException: Content.Server._CM14.Headbutt.HeadbuttSystem requested unregistered type with its field Content.Shared._CM14.Xenos.Animations.SharedXenoAnimationsSystem: _xenoAnimation
```

File: Resources/Prototypes/_CM14/Actions/Xeno/xeno_offense_actions.yml (hunk 6 above)
Comment (DrSmugleaf) [reply Resources/Prototypes/_CM14/Actions/Xeno/xeno_offense_actions.yml]: Tested this too, seems to work reliably

**Review by DrSmugleaf — APPROVED** [review]
Thanks for the PR!

**Pull request merged**